proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,469 @@
+                self.update(|metrics| {
+                    *metrics.request_total(&labels) += 1;
+                })
+            },
+
+            Event::StreamRequestFail(_, _) => {
+                // The request total was already incremented by the
+                // StreamRequestOpen event; when we count requests on
+                // stream ends, we'll care about this event.
+            },
+
+            Event::StreamResponseOpen(_, _) =>  {
+                // Wait until the response stream ends to record it.
+            },
+
+            Event::StreamResponseEnd(ref res, ref end) => {
+                let labels = Arc::new(ResponseLabels::new(
+                    res,
+                    end.grpc_status,
+                ));
+                self.update(|metrics| {
+                    *metrics.response_total(&labels) += 1;
+                    *metrics.response_duration(&labels) +=  end.since_response_open;
+                    *metrics.response_latency(&labels) += end.since_request_open;
+                });
+            },
+
+            Event::StreamResponseFail(ref res, ref fail) => {
+                // TODO: do we care about the failure's error code here?
+                let labels = Arc::new(ResponseLabels::new(res, None));
+                self.update(|metrics| {
+                    *metrics.response_total(&labels) += 1;
+                    *metrics.response_duration(&labels) += fail.since_response_open;
+                    *metrics.response_latency(&labels) += fail.since_request_open;
+                });
+            },
+
+            Event::TransportOpen(_) | Event::TransportClose(_, _) => {
+                // TODO: we don't collect any metrics around transport events.
+            },
+        };
+    }
+}
+
+
+// ===== impl Serve =====
+
+impl Serve {
+    fn new(metrics: &Arc<Mutex<Metrics>>) -> Self {
+        Serve { metrics: metrics.clone() }
+    }
+}
+
+impl HyperService for Serve {
+    type Request = HyperRequest;
+    type Response = HyperResponse;
+    type Error = hyper::Error;
+    type Future = FutureResult<Self::Response, Self::Error>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+        if req.path() != "/metrics" {
+            return future::ok(HyperResponse::new()
+                .with_status(StatusCode::NotFound));
+        }
+
+        let body = {
+            let metrics = self.metrics.lock()
+                .expect("metrics lock poisoned");
+            format!("{}", *metrics)
+        };
+        future::ok(HyperResponse::new()
+            .with_header(ContentLength(body.len() as u64))
+            .with_header(ContentType::plaintext())
+            .with_body(body))
+    }
+}
+
+
+// ===== impl RequestLabels =====
+
+impl<'a> RequestLabels {
+    fn new(req: &ctx::http::Request) -> Self {
+        let outbound_labels = if req.server.proxy.is_outbound() {
+            Some(OutboundLabels {
+                // TODO: when #429 is done, add appropriate destination label.
```

Comment:
So I understand, is this the answer to the question  "do any of the OutboundLabels get set?"?

Reply:
Yeah, that's correct. The proxy doesn't currently have access to this information. #429 should add the label value to the control plane's `Destination` response.